tests/integration/io/test_list_files_gcs.py

```diff
@@ -0,0 +1,45 @@
+from __future__ import annotations
+
+import gcsfs
+import pytest
+
+from daft.daft import io_list
+
+BUCKET = "daft-public-data-gs"
+
+
+def compare_gcs_result(daft_ls_result: list, fsspec_result: list):
+    daft_files = [(f["path"], f["type"].lower()) for f in daft_ls_result]
+    gcsfs_files = [(f"gs://{f['name']}", f["type"]) for f in fsspec_result]
+
+    # Perform necessary post-processing of fsspec results to match expected behavior from Daft:
+
+    # NOTE: gcsfs sometimes does not return the trailing / for directories, so we have to ensure it
+    gcsfs_files = [
+        (f"{path.rstrip('/')}/", type_) if type_ == "directory" else (path, type_) for path, type_ in gcsfs_files
+    ]
+
+    # NOTE: gcsfs will sometimes return 0-sized marker files for manually-created folders, which we ignore here
+    # Be careful here because this will end up pruning any truly size-0 files that are actually files and not folders!
+    size_0_files = {f"gs://{f['name']}" for f in fsspec_result if f["size"] == 0 and f["type"] == "file"}
+    gcsfs_files = [(path, type_) for path, type_ in gcsfs_files if path not in size_0_files]
+
+    assert sorted(daft_files) == sorted(gcsfs_files)
+
+
+@pytest.mark.integration()
+@pytest.mark.parametrize(
+    "path",
+    [
+        f"gs://{BUCKET}",
+        f"gs://{BUCKET}/",
+        f"gs://{BUCKET}/test_ls",
+        f"gs://{BUCKET}/test_ls/",
+        f"gs://{BUCKET}/test_ls/paginated-1100-files/",
+    ],
+)
+def test_gs_flat_directory_listing(path):
```

Comment:
can you also add the recursive version? `io_list(recursive=True)`. I didn't finish writing all my tests yet for s3

Reply:
Yes, what's the intended behavior for `io_list(recursive=True)` for intermediate folders?

It seems we have a discrepancy in behavior where `io_list` prunes intermediate `"directory"` entries, but fsspec `glob(.../**)` will return intermediate directory entries.

Reply:
Fixed: https://github.com/Eventual-Inc/Daft/pull/1392/commits/401ca8eebb563ce60b2bdbcabef2e2b0b3b4dc34